### PR TITLE
KnownReference: Add support for testing for referenced libraries 

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
A precondition for #6662

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        internal KnownReference(Func<AssemblyIdentity, bool> predicate) : this(Any(predicate))
+        {
+        }
+
+        internal KnownReference(Func<IEnumerable<AssemblyIdentity>, bool> predicate)
+        {
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
+        }
+
+        public static KnownReference XUnit_Assert { get; } = new(NameIs("xunit.assert").Or(NameIs("xunit").And(VersionLowerThen("2.0"))));
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:36]: I created a small DSL to define the references. This allows for unifying the checks which should improve maintainability (e.g. the Name, Version, and PublicKey checks might be buggy).

XUnit extracted the "Assert" part in version 2 into a separate assembly so the check is
* "xunit.assert" or
* "xunit" and Version before 2.0

Something like this could also be used for e.g. "SqlClient" for Microsoft SQL server which is distributed in several different assemblies [System.Data.SqlClient](https://www.nuget.org/packages/System.Data.SqlClient) and [Microsoft.Data.SqlClient](https://www.nuget.org/packages/Microsoft.Data.SqlClient)

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers;
+
+public sealed partial class KnownReference
+{
+    internal static Func<AssemblyIdentity, bool> NameIs(string name) =>
+        x => x.Name.Equals(name);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:26]: I am not sure about the casing of assembly names. Currently, I use the standard string.Equals but this is most likely wrong:

> The runtime treats assembly names as case-insensitive when binding to an assembly, but preserves whatever case is used in an assembly name. Several tools in the Windows SDK handle assembly names as case-sensitive. For best results, manage assembly names as though they were case-sensitive.

[Source](https://learn.microsoft.com/en-us/dotnet/standard/assembly/names)

As far as I know, you can only load assemblies in the run-time with unique names, and the casing is ignored. I think it would be better to do `StringComparison.OrdinalIgnoreCase` but I'm not sure.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        x => x.Version >= version;
+
+    internal static Func<AssemblyIdentity, bool> VersionBetween(string from, string to) =>
+        VersionBetween(Version.Parse(from), Version.Parse(to));
+
+    internal static Func<AssemblyIdentity, bool> VersionBetween(Version from, Version to) =>
+        x => x.Version >= from && x.Version <= to;
+
+    internal static Func<AssemblyIdentity, bool> OptionalPublicKeyTokenIs(string key) =>
+        x => !x.HasPublicKey || PublicKeyEqualHex(x, key);
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:56]: for cases like [StackExchange.Redis](https://www.nuget.org/packages?q=StackExchange.Redis) which comes with a strong name or without.

File: analyzers/src/SonarAnalyzer.Common/Helpers/CompilationExtensions.cs [hunk 4]
```diff
@@ -31,5 +31,8 @@ public static class CompilationExtensions
             // There's no direct way of checking compilation target framework yet (09/2020).
             // See https://github.com/dotnet/roslyn/issues/3798
             compilation.ObjectType.ContainingAssembly.Name == "mscorlib";
+
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/CompilationExtensions.cs:34]: While touching this file, you can move it to "Extensions" directory and change the namespace

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:26]: Let's play safe and use case insensitive. I can't imagine that xUnit 1.0 would be something else than xunit 1.0

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+public sealed partial class KnownReference
+{
+    internal static Func<AssemblyIdentity, bool> NameIs(string name) =>
+        x => x.Name.Equals(name);
+
+    internal static Func<AssemblyIdentity, bool> StartsWith(string name) =>
+        x => x.Name.StartsWith(name);
+
+    internal static Func<AssemblyIdentity, bool> EndsWith(string name) =>
+        x => x.Name.EndsWith(name);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:32]: Do you have a clear example in mind? Experience with KnownType shows that we typically exactly know what we're looking for.
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:56]: I don't think we'll ever need to be that precise.
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        VersionBetween(Version.Parse(from), Version.Parse(to));
+
+    internal static Func<AssemblyIdentity, bool> VersionBetween(Version from, Version to) =>
+        x => x.Version >= from && x.Version <= to;
+
+    internal static Func<AssemblyIdentity, bool> OptionalPublicKeyTokenIs(string key) =>
+        x => !x.HasPublicKey || PublicKeyEqualHex(x, key);
+
+    internal static Func<AssemblyIdentity, bool> PublicKeyTokenIs(string key) =>
+        x => x.HasPublicKey && PublicKeyEqualHex(x, key);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:59]: Same as above
```suggestion
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers;
+
+public sealed partial class KnownReference
+{
+    internal static Func<AssemblyIdentity, bool> NameIs(string name) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:25]: Does this need to be internal? For testing purposes, add more static fields with some well-known libraries (like the xunit) that will use the feature.
For some hell-versions, take a look at `TestMethodShouldContainAssertionTest` for inspiration
```suggestion
    private static Func<AssemblyIdentity, bool> NameIs(string name) =>
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace SonarAnalyzer.Helpers
+{
+    public sealed partial class KnownReference
+    {
+        private readonly Func<IEnumerable<AssemblyIdentity>, bool> predicate;
+
+        internal KnownReference(Func<AssemblyIdentity, bool> predicate) : this(Any(predicate))
+        {
+        }
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:29]: Can be like this
```suggestion
        internal KnownReference(Func<AssemblyIdentity, bool> predicate) : this(Any(predicate)) { }
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+        internal KnownReference(Func<IEnumerable<AssemblyIdentity>, bool> predicate)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:31]: Arrow constructor

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:36]: Static fields should be above constructors

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs (hunk 9 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:31]: private should be enough, shouldn't it?

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:36]: It's nice, but somewhat hard to follow with all the possibilities. We'd typically need A || (B && C) || (D && E && F) schema.
As a first step, let's remove `Or`, keep `And`, and make it `params` that will behave as "any". That should simplify it.

```suggestion
        public static KnownReference XUnit_Assert { get; } = new(
            NameIs("xunit.assert"), 
            NameIs("xunit").And(VersionLowerThen("2.0"))));
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    private static bool PublicKeyEqualHex(AssemblyIdentity identity, string hexString)
+    {
+        var normalizedHexString = hexString.Replace("-", string.Empty);
+        return ArraysEqual(identity.PublicKeyToken.ToArray(), normalizedHexString) || ArraysEqual(identity.PublicKey.ToArray(), normalizedHexString);
+
+        static bool ArraysEqual(byte[] key, string hexString) =>
+            BitConverter.ToString(key).Replace("-", string.Empty).Equals(hexString, StringComparison.OrdinalIgnoreCase);
+    }
+
+    internal static Func<IEnumerable<AssemblyIdentity>, bool> Any(Func<AssemblyIdentity, bool> predicate) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:70]: This will probably go directly into the constructor after other comments

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReferenceExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace SonarAnalyzer.Helpers;
+
+internal static class KnownReferenceExtensions
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReferenceExtensions.cs:23]: Should be in `Extensions` directory and namespace

**Review by pavel-mikula-sonarsource — CHANGES_REQUESTED** [review]
First round, I didn't look into the UTs yet as things will change a bit. I asked for some simplifications

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:32]: For StartsWith I was thinking of something like [Azure.Storage](https://www.nuget.org/packages?q=Azure.Storage&packagetype=&frameworks=&tfms=&prerel=True&sortby=relevance), [Azure.ResourceManager](https://www.nuget.org/packages?q=Azure.ResourceManager&packagetype=&frameworks=&tfms=&prerel=True&sortby=relevance), [AWSSDK](https://www.nuget.org/packages?q=AWSSDK+&packagetype=&frameworks=&tfms=&prerel=True&sortby=relevance) or [Microsoft.Extensions](https://www.nuget.org/packages?q=Microsoft.Extensions&packagetype=&frameworks=&tfms=&prerel=True&sortby=relevance) and for Contains something like [IdentityModel.Tokens](https://www.nuget.org/packages?q=IdentityModel.Tokens) or [SqlClient](https://www.nuget.org/packages?q=SqlClient&packagetype=&frameworks=&tfms=&prerel=True&sortby=relevance) or [Logger](https://www.nuget.org/packages?q=Logger&packagetype=&frameworks=&tfms=&prerel=True&sortby=relevance)

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 6 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:59]: If there is a strong name it is the most reliable source of identity. Also if we don't add the possibility to check for the strong name no one will ever add the functionality even in cases where it would be useful. The functionality is now there and fully covered by tests with our own public key.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 7 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs:25]: I'd rather like to test the single elements of the "DSL" for better test isolation. Relying on `public static KnownReference XUnit_Assert { get; }` and the like is problematic as these definitions change over time and any tests that rely on it will deteriorate. I did this to improve the situation

* Encapsulate the methods in a nested internal `Predicates` class
* Mark the class with `[EditorBrowsable(EditorBrowsableState.Never)]` to suppress IntelliSense from showing the nested class as much as possible.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs (hunk 9 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs]: > private should be enough, shouldn't it?

I have some tests that call the constructors.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs]: We don't need to be that precise at all. We'll likely use these for an early bailout, as you plan to use them. And there will be no harm if we trigger on something that is a false match.

Let's remove it.

Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs]: I also highly doubt that we would ever track a library that updates it's public key between versions.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs]: The `EditorBrowsable` is too strong solution. The nested class will do the trick.

File: analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs [hunk 12]
```diff
@@ -24,4 +24,17 @@ internal static class CompilationExtensions
 {
     public static INamedTypeSymbol GetTypeByMetadataName(this Compilation compilation, KnownType knownType) =>
         compilation.GetTypeByMetadataName(knownType.FullName);
+
+    public static IMethodSymbol GetTypeMethod(this Compilation compilation, SpecialType type, string methodName) =>
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs:28]:
```suggestion
    public static IMethodSymbol SpecialTypeMethod(this Compilation compilation, SpecialType type, string methodName) =>
```

File: analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs [hunk 13]
```diff
@@ -24,4 +24,17 @@ internal static class CompilationExtensions
...
+        (IMethodSymbol)compilation.GetSpecialType(type)
+            .GetMembers(methodName)
+            .SingleOrDefault();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs:31]: This could be a single line

File: analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs [hunk 14]
```diff
@@ -24,4 +24,17 @@ internal static class CompilationExtensions
...
+
+    public static bool IsNetFrameworkTarget(this Compilation compilation) =>
+        // There's no direct way of checking compilation target framework yet (09/2020).
+        // See https://github.com/dotnet/roslyn/issues/3798
+        compilation.ObjectType.ContainingAssembly.Name == "mscorlib";
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs:36]: This is somewhat overlapping with `NetFrameworkVersionProvider` class
https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.Common/Helpers/NetFrameworkVersionProvider.cs

While this extension can live here, the implementation should be in that class to have this kind of ugly detection should live on one place.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+{
+    public sealed partial class KnownReference
+    {
+        private readonly Func<IEnumerable<AssemblyIdentity>, bool> predicate;
+
+        public static KnownReference XUnit_Assert { get; } = new(
+            NameIs("xunit.assert"),
+            NameIs("xunit").And(VersionLowerThen("2.0")));
+
+        internal KnownReference(Func<AssemblyIdentity, bool> predicate, params Func<AssemblyIdentity, bool>[] or)
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:33]: This will simplify it quite a lot. As this is supposed to be instantiated only from within this class (and therefore it would be beneficial to have it `private`), we don't need to enforce it syntactically. You can throw in case it's empty
```suggestion
        internal KnownReference(params Func<AssemblyIdentity, bool>[] orPredicates)
```

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using static SonarAnalyzer.Helpers.KnownReference.Predicates;
+
+namespace SonarAnalyzer.Helpers
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:23]: Could be file scoped namespace

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+
+        internal KnownReference(Func<AssemblyIdentity, bool> predicate, params Func<AssemblyIdentity, bool>[] or)
+            : this(predicate is null || or.Any(x => x is null)
+                  ? throw new ArgumentNullException(nameof(predicate), "All predicates must be non-null.")
+                  : identities => identities.Any(identitiy => predicate(identitiy) || or.Any(orPredicate => orPredicate(identitiy))))
+        {
+        }
+
+        internal KnownReference(Func<IEnumerable<AssemblyIdentity>, bool> predicate) =>
+            this.predicate = predicate ?? throw new ArgumentNullException(nameof(predicate));
```
Comment (pavel-mikula-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs:41]: We don't need two constructors, this class is meant to be created only with the previous one. Assign the predicate there. Those few UTs will survive without it ,they can use the other one anyway
```suggestion
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using Moq;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Extensions;
+using static SonarAnalyzer.Helpers.KnownReference;
+using static SonarAnalyzer.Helpers.KnownReference.Predicates;
+
+namespace SonarAnalyzer.UnitTest;
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs:27]: Thsi file should live in `Helpers` directory, same as the class it tests.
And have `.Helpers` namespace
```suggestion
namespace SonarAnalyzer.UnitTest.Helpers;
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+    [DataRow("MyTestMy", false)]
+    [DataRow("MyTESTMy", false)]
+    [DataRow("Without", false)]
+    public void NameIs_Test(string name, bool expected)
+    {
+        var sut = new KnownReference(NameIs("Test"));
+        var compilation = new Mock<Compilation>("compilationName", ImmutableArray<MetadataReference>.Empty, new Dictionary<string, string>(), false, null, null);
+        var identity = new AssemblyIdentity(name);
+        compilation.SetupGet(x => x.ReferencedAssemblyNames).Returns(new[] { identity });
+        sut.IsReferencedBy(compilation.Object).Should().Be(expected);
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs:68]: Let's extract the repetitive scaffodling into a static method
```suggestion
        sut.IsReferencedBy(CreateCompilation(name)).Should().Be(expected);
```

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs:68]: Version can be an optional parameter

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    [TestMethod]
+    public void XUnitAssert_2_4()
+    {
+        var compilation = SolutionBuilder.Create()
+            .AddProject(AnalyzerLanguage.CSharp)
+            .AddReferences(NuGetMetadataReference.XunitFramework("2.4.2"))
+            .AddSnippet("// Empty file")
+            .GetCompilation();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs:271]: There's a helper for this
```suggestion
        var compilation = TestHelper.CompileCS("// Empty file", NuGetMetadataReference.XunitFramework("2.4.2")).Model.Compilation
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+    }
+
+    [TestMethod]
+    public void XUnitAssert_1_9()
+    {
+        var compilation = SolutionBuilder.Create()
+            .AddProject(AnalyzerLanguage.CSharp)
+            .AddReferences(NuGetMetadataReference.XunitFrameworkV1)
+            .AddSnippet("// Empty file")
+            .GetCompilation();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs:282]: Same here

File: analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs [hunk 22]
```diff
@@ -0,0 +1,295 @@
...
+        compilation.References(XUnit_Assert).Should().BeTrue();
+    }
+
+    [TestMethod]
+    public void XUnitAssert_NoReference()
+    {
+        var compilation = SolutionBuilder.Create()
+            .AddProject(AnalyzerLanguage.CSharp)
+            .AddSnippet("// Empty file")
+            .GetCompilation();
```
Comment (pavel-mikula-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Extensions/KnownReferenceTests.cs:292]: Also here

**Review by pavel-mikula-sonarsource — APPROVED** [review]
LGTM to unblock the sprint.
I left further improvement comments.
Please also remove the methods from the previous review. We won't need them any time soon.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs (hunk 5 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.Predicates.cs]: There are also checks like "IsTestProject" or "Is this targeting .Net Core" or "Is this a WebProject" or "Is this a cloud app". These shortcuts can come in handy for those.

File: analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs (hunk 14 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Extensions/CompilationExtensions.cs:36]: The implementations can not easily be merged. I created https://github.com/SonarSource/sonar-dotnet/issues/6751

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs (hunk 15 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs]: This would be a bad API design. We can enforce proper usage at compile time and should do so.

File: analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs (hunk 17 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Helpers/KnownReference.cs]: This ctor is needed for cases where more than one assembly reference needs to be present (e.g. WPF: WindowsBase, PresentationCore, and PresentationFramework).
